### PR TITLE
Improve logs and add timeouts

### PR DESCRIPTION
**Pull request by ayalash**

### CODE REVIEW

File: k8s-agent/src/k8s/pods.rs [hunk 1]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
         });
 
     match result {
-        Ok(exe_result) => Ok(exe_result.output),
-        Err(exe_result) => Err(exe_result),
+        Ok(exe_result) => {
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:54]:
```
result.map(|exe_result| {
   debug!(...);
   exe_result.output
}.tap_err(...)

File: k8s-agent/src/k8s/pods.rs [hunk 2]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+        }
+        Err(exe_result) => {
+            debug!("{} deletion failed", pod_name);
+            Err(exe_result)
+        }
     }
 }
 
-async fn wait_for_pod_phase<S, F>(mut stream: S, pred: F) -> Result<()>
+async fn create_interval_stream(duration: Duration) -> IntervalStream {
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:65]: Doesn't have to be `async`

File: k8s-agent/src/k8s/pods.rs [hunk 3]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
-                    let status = match pod.status.as_ref() {
-                        Some(status) => status,
-                        None => continue,
-                    };
-                    match &status.phase {
-                        Some(phase) => {
-                            if pred(phase) {
-                                return Ok(());
+    let mut logs_timer_stream = create_interval_stream(Duration::from_secs(60)).await;
+    let mut timeout_timer_stream = create_interval_stream(timeout_duration).await;
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:82]: I think this should be a `sleep`

File: k8s-agent/src/k8s/pods.rs [hunk 4]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
-        .await
-        .context("Failed waiting for Helm pod to reach Running phase")?;
-    info!("Ready to attach to {}", pod_name);
-
-    let attached = pods.attach(pod_name, &Default::default()).await?;
-    let output = get_pod_output(attached).await?;
-
-    wait_for_pod_phase(&mut pod_events, |p| p == "Succeeded" || p == "Failed")
+    let is_pod_finished =
+        |phase_name: &str| -> bool { phase_name == "Succeeded" || phase_name == "Failed" };
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:127]: I would change to eq_ignore_ascii_case

File: k8s-agent/src/k8s/pods.rs [hunk 5]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+
+    let mut pod_phase = wait_for_pod_phase(
+        &mut pod_events,
+        |p| p == "Running" || is_pod_finished(p),
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:131]: Same here

File: k8s-agent/src/k8s/pods.rs [hunk 6]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+        Duration::from_secs(60),
+    )
+    .await
+    .with_context(|| {
+        format!(
+            "Failed waiting for Helm pod {} to reach Running phase",
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:137]: Interpolate variable here

File: k8s-agent/src/k8s/pods.rs [hunk 7]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+            pod_name
+        )
+    })?;
+    info!("Ready to attach to {} (phase: {})", pod_name, pod_phase);
+
+    let output = if is_pod_finished(&pod_phase) {
+        "<N/A>".into()
+    } else {
+        debug!("Attaching to pod {}", pod_name);
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:146]: Interpolate variable

File: k8s-agent/src/k8s/pods.rs [hunk 8]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
 
     match result {
-        Ok(exe_result) => Ok(exe_result.output),
-        Err(exe_result) => Err(exe_result),
+        Ok(exe_result) => {
+            debug!("{} deletion succeed", pod_name);
+            Ok(exe_result.output)
+        }
+        Err(exe_result) => {
+            debug!("{} deletion failed", pod_name);
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:59]: Interpolate variables

File: k8s-agent/src/k8s/pods.rs [hunk 9]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
         });
 
     match result {
-        Ok(exe_result) => Ok(exe_result.output),
-        Err(exe_result) => Err(exe_result),
+        Ok(exe_result) => {
+            debug!("{} deletion succeed", pod_name);
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:55]: Interpolate variables

File: k8s-agent/src/k8s/pods.rs [hunk 10]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+            &mut pod_events,
+            is_pod_finished,
+            Duration::from_secs(60 * 10),
+        )
         .await
-        .context("Failed waiting for Helm pod to reach Succeeded or Failed phase")?;
-    info!("Pod {} terminated", pod_name);
+        .with_context(|| {
+            format!(
+                "Failed waiting for Helm pod {} to reach Succeeded or Failed phase",
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:158]: Interpolate variables

File: k8s-agent/src/k8s/pods.rs [hunk 11]
```diff
@@ -47,61 +51,117 @@ pub async fn execute_pod(pods: Api<Pod>, pod: Pod) -> Result<String> {
...
+                pod_name
+            )
+        })?;
+        output
+    };
+    info!("Pod {} terminated (phase: {})", pod_name, pod_phase);
```
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:164]: Interpolate variables

File: k8s-agent/src/k8s/pods.rs (hunk 5 above)
Comment (vmalloc) [inline k8s-agent/src/k8s/pods.rs:131]: Why not unite these three conditions? Maybe call it `pod_not_pending`?

**Review by vmalloc — CHANGES_REQUESTED** [review]

**Comment by ayalash** [comment]
@vmalloc Fixed.
Please review

**Review by popen2 — APPROVED** [review]

**Pull request merged**